calico/felix/fiptables.py

```diff
@@ -290,31 +299,83 @@ def delete_chains(self, table_name, chain_names, callback=None):
         if callback:
             self.completion_callbacks.append(callback)
 
+    @actor_event
+    def cleanup(self):
+        """
+        Trigger clean up
+        """
+        # In case someone deleted something from under us, reload from the
+        # dataplane.  This has side-effects outside our batch but only this
+        # method cares and it's important that we start with fresh information.
+        self._load_from_dataplane()
+        for table, chains in self.chains_in_dataplane.iteritems():
+            orphan_chains = chains - self.explicitly_prog_chains[table]
+            # Filter out chains that are already touched by this batch.  Note:
+            # We do not try to filter out chains that are referenced but not
+            # explicitly programmed, we'll catch those in _finish_msg_batch()
+            # and reprogram them as a stub.
+            chains_to_delete = [c for c in orphan_chains
+                                if c not in self.bch_affected_chains[table] and
+                                   c.startswith(FELIX_PREFIX)]
+            # SMC: It'd be nice if we could do a best-effort delete on these
+            # chains but that's hard to do since they'll all be processed as
+            # one atomic iptables-restore.
+            self.delete_chains(table, chains_to_delete)
+
     def _start_msg_batch(self, batch):
         self._reset_batched_work()
         return batch
 
     def _finish_msg_batch(self, batch, results):
         start = time.time()
-
+        modify_succeeded = False
         try:
             self._calculate_index_changes()
-            input_lines = self._calculate_ipt_input()
+            input_lines = self._calculate_ipt_modify_input()
             self._execute_iptables(input_lines)
+            modify_succeeded = True
+            try:
+                input_lines = self._calculate_ipt_delete_input()
```

Comment:
I split the iptables calls into two batches, one that modifies the chains (and replaces any chains that we're deleting with a drop rule).  And then one that actually does the delete.  This two-pass approach means that:

* we can better handle failure to delete chains that are referenced by orphaned chains that we haven't cleaned up yet
* when we're deleting orphan chains, we decouple them all so they don't reference each other before we actually do the delete
* we fail safe, dropping traffic rather than failing to delete a chain.